### ISSUE TITLE
The following errors were encountered during discovery: ifPhysAddress

### ISSUE THREAD

**Issue by netravnen** [renatoalmeidaoliveira/snmp-diode#2]
**Description**

I am trying to run a discovery against localhost (LXC Container on [PVE](https://www.proxmox.com/en/)) and Mikrotik CCR and CRS devices (RouterOS 7.16.1).

I am hitting the same error.

_Could_ this be related to the blank MAC Address returned in the SNMP data for the linux `lo` interface?
- https://github.com/renatoalmeidaoliveira/snmp-diode/blob/main/snmp_diode/entrypoint.py#L130
- https://github.com/renatoalmeidaoliveira/snmp-diode/blob/main/snmp_diode/discover.py#L66-L73

**Error**

snmp-diode against localhost

```
root@diode1 ~ # /opt/snmp-diode/venv/bin/snmp-diode \
  --host localhost --version 2 --community public \
  --diode grpc://localhost:8081/diode --api_key REDACTED

ERROR: The following errors were encountered during discovery:
ERROR: localhost - 'ifPhysAddress'
[]
```

```
# uname -a
Linux diode1 6.8.4-2-pve #1 SMP PREEMPT_DYNAMIC PMX 6.8.4-2 (2024-04-10T17:36Z) x86_64 GNU/Linux
```

```
# lsb_release -d
Description:    Debian GNU/Linux 12 (bookworm)
```

```
# systemd-detect-virt 
lxc
```

**Bulk Walk**

```
# echo "\n### SNMP OID 1.3.6.1.2.1.1.5.0       ###\n" && /bin/snmpwalk -v 2c -c public localhost 1.3.6.1.2.1.1.5.0       && \
  echo "\n### SNMP OID 1.3.6.1.2.1.1.2.0       ###\n" && /bin/snmpwalk -v 2c -c public localhost 1.3.6.1.2.1.1.2.0       && \
  echo "\n### SNMP OID 1.3.6.1.2.1.1.6.0       ###\n" && /bin/snmpwalk -v 2c -c public localhost 1.3.6.1.2.1.1.6.0       && \
  echo "\n### SNMP OID 1.3.6.1.2.1.2.2.1.2     ###\n" && /bin/snmpwalk -v 2c -c public localhost 1.3.6.1.2.1.2.2.1.2     && \
  echo "\n### SNMP OID 1.3.6.1.2.1.31.1.1.1.18 ###\n" && /bin/snmpwalk -v 2c -c public localhost 1.3.6.1.2.1.31.1.1.1.18 && \
  echo "\n### SNMP OID 1.3.6.1.2.1.2.2.1.6     ###\n" && /bin/snmpwalk -v 2c -c public localhost 1.3.6.1.2.1.2.2.1.6     && \
  echo "\n### SNMP OID 1.3.6.1.2.1.2.2.1.7     ###\n" && /bin/snmpwalk -v 2c -c public localhost 1.3.6.1.2.1.2.2.1.7     && \
  echo "\n### SNMP OID 1.3.6.1.2.1.4.20.1.2    ###\n" && /bin/snmpwalk -v 2c -c public localhost 1.3.6.1.2.1.4.20.1.2    && \
  echo "\n### SNMP OID 1.3.6.1.2.1.4.20.1.3    ###\n" && /bin/snmpwalk -v 2c -c public localhost 1.3.6.1.2.1.4.20.1.3
```

Output

```
### SNMP OID 1.3.6.1.2.1.1.5.0       ###

SNMPv2-MIB::sysName.0 = STRING: diode1

### SNMP OID 1.3.6.1.2.1.1.2.0       ###

SNMPv2-MIB::sysObjectID.0 = OID: NET-SNMP-MIB::netSnmpAgentOIDs.10

### SNMP OID 1.3.6.1.2.1.1.6.0       ###

SNMPv2-MIB::sysLocation.0 = STRING: REDACTED

### SNMP OID 1.3.6.1.2.1.2.2.1.2     ###

IF-MIB::ifDescr.1 = STRING: lo
IF-MIB::ifDescr.2 = STRING: eth0
IF-MIB::ifDescr.3 = STRING: br-9ae2448c1c2b
IF-MIB::ifDescr.4 = STRING: docker0
IF-MIB::ifDescr.110 = STRING: veth87c9a11
IF-MIB::ifDescr.114 = STRING: veth9a8f3f0
IF-MIB::ifDescr.116 = STRING: vethcb6ed93
IF-MIB::ifDescr.118 = STRING: veth0e5d8d0

### SNMP OID 1.3.6.1.2.1.31.1.1.1.18 ###

IF-MIB::ifAlias.1 = STRING: 
IF-MIB::ifAlias.2 = STRING: 
IF-MIB::ifAlias.3 = STRING: 
IF-MIB::ifAlias.4 = STRING: 
IF-MIB::ifAlias.110 = STRING: 
IF-MIB::ifAlias.114 = STRING: 
IF-MIB::ifAlias.116 = STRING: 
IF-MIB::ifAlias.118 = STRING: 

### SNMP OID 1.3.6.1.2.1.2.2.1.6     ###

IF-MIB::ifPhysAddress.1 = STRING: 
IF-MIB::ifPhysAddress.2 = STRING: bc:24:11:86:96:81
IF-MIB::ifPhysAddress.3 = STRING: 2:42:7e:4c:c8:d7
IF-MIB::ifPhysAddress.4 = STRING: 2:42:6d:24:10:c5
IF-MIB::ifPhysAddress.110 = STRING: 86:d:c2:14:e5:6d
IF-MIB::ifPhysAddress.114 = STRING: f6:ea:ee:40:dc:18
IF-MIB::ifPhysAddress.116 = STRING: c6:14:36:4b:62:c9
IF-MIB::ifPhysAddress.118 = STRING: 52:f3:44:b6:a2:d8

### SNMP OID 1.3.6.1.2.1.2.2.1.7     ###

IF-MIB::ifAdminStatus.1 = INTEGER: up(1)
IF-MIB::ifAdminStatus.2 = INTEGER: up(1)
IF-MIB::ifAdminStatus.3 = INTEGER: up(1)
IF-MIB::ifAdminStatus.4 = INTEGER: up(1)
IF-MIB::ifAdminStatus.110 = INTEGER: up(1)
IF-MIB::ifAdminStatus.114 = INTEGER: up(1)
IF-MIB::ifAdminStatus.116 = INTEGER: up(1)
IF-MIB::ifAdminStatus.118 = INTEGER: up(1)

### SNMP OID 1.3.6.1.2.1.4.20.1.2    ###

IP-MIB::ipAdEntIfIndex.127.0.0.1 = INTEGER: 1
IP-MIB::ipAdEntIfIndex.172.17.0.1 = INTEGER: 4
IP-MIB::ipAdEntIfIndex.172.18.0.1 = INTEGER: 3
IP-MIB::ipAdEntIfIndex.192.168.192.75 = INTEGER: 2

### SNMP OID 1.3.6.1.2.1.4.20.1.3    ###

IP-MIB::ipAdEntNetMask.127.0.0.1 = IpAddress: 255.0.0.0
IP-MIB::ipAdEntNetMask.172.17.0.1 = IpAddress: 255.255.0.0
IP-MIB::ipAdEntNetMask.172.18.0.1 = IpAddress: 255.255.0.0
IP-MIB::ipAdEntNetMask.192.168.192.75 = IpAddress: 255.255.255.248
```

**Comment by netravnen** [comment]
```python
# /opt/snmp-diode/venv/bin/python3
Python 3.11.2 (main, Sep 14 2024, 03:00:30) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from easysnmp import snmp_get, snmp_set, snmp_walk, Session
>>> session = Session(hostname='localhost', community='public', version=2)
>>> session.walk('1.3.6.1.2.1.2.2.1.6')
[<SNMPVariable value='' (oid='ifPhysAddress', oid_index='1', snmp_type='OCTETSTR')>, <SNMPVariable value='$ (contains binary)' (oid='ifPhysAddress', oid_index='2', snmp_type='OCTETSTR')>, <SNMPVariable value='B~L (contains binary)' (oid='ifPhysAddress', oid_index='3', snmp_type='OCTETSTR')>, <SNMPVariable value='Bm$ (contains binary)' (oid='ifPhysAddress', oid_index='4', snmp_type='OCTETSTR')>, <SNMPVariable value='\rm (contains binary)' (oid='ifPhysAddress', oid_index='110', snmp_type='OCTETSTR')>, <SNMPVariable value='@ (contains binary)' (oid='ifPhysAddress', oid_index='114', snmp_type='OCTETSTR')>, <SNMPVariable value='6Kb (contains binary)' (oid='ifPhysAddress', oid_index='116', snmp_type='OCTETSTR')>, <SNMPVariable value='RD (contains binary)' (oid='ifPhysAddress', oid_index='118', snmp_type='OCTETSTR')>]
```

**Comment by netravnen** [comment]
My problem was the docker container was outdated. Found the below thread in Slack. The resolution was to update the docker container used locally.

https://netdev-community.slack.com/archives/C07DHA9G0HL/p1729690344866149

```
# docker image ls | grep netbox
netboxlabs/diode-reconciler   latest    0fe13694be73   2 months ago    44.1MB
netboxlabs/diode-ingester     latest    eceb9f3f7ce0   2 months ago    42MB
netboxlabs/diode-reconciler   <none>    bf7bb52ce8df   4 months ago    42.2MB
netboxlabs/diode-ingester     <none>    02201071a609   4 months ago    41.6MB
```

```
docker pull netboxlabs/diode-reconciler:latest && \
docker pull netboxlabs/diode-ingester:latest && \
docker pull redis/redis-stack-server:latest && \
docker pull nginx:latest
```

```
docker image prune --all
```

**Comment by netravnen** [comment]
... Not really 😕 

```
# /opt/snmp-diode/venv/bin/snmp-diode --host localhost --version 2 \
  --community public --diode grpc://localhost:8080/diode --api_key REDACTED

ERROR: The following errors were encountered during discovery:
ERROR: localhost - 'ifPhysAddress'
```

**Comment by tabasco0322** [comment]
Having the same issue. 
Bunch of blank MAC address on loads of interfaces.

**Comment by cguadall** [comment]
Same error here running against Cisco Small Business switches

$ /usr/local/bin/snmp-diode -n redacted-ip -v 2 -c redacted-community
ERROR: The following errors were encountered during discovery:
ERROR: redacted-ip - 'ifPhysAddress'

Checking for the ifPhysAddress table we can see that there are severeal entries with the same mac address

$ snmpwalk -v 2c -c redacted-community redacted-ip .1.3.6.1.2.1.2.2.1.6
IF-MIB::ifPhysAddress.1 = STRING: a0:f8:49:f6:9e:56
IF-MIB::ifPhysAddress.2 = STRING: a0:f8:49:f6:9e:57
IF-MIB::ifPhysAddress.3 = STRING: a0:f8:49:f6:9e:58
IF-MIB::ifPhysAddress.4 = STRING: a0:f8:49:f6:9e:59
IF-MIB::ifPhysAddress.5 = STRING: a0:f8:49:f6:9e:5a
IF-MIB::ifPhysAddress.6 = STRING: a0:f8:49:f6:9e:5b
IF-MIB::ifPhysAddress.7 = STRING: a0:f8:49:f6:9e:5c
IF-MIB::ifPhysAddress.8 = STRING: a0:f8:49:f6:9e:5d
IF-MIB::ifPhysAddress.9 = STRING: a0:f8:49:f6:9e:5e
IF-MIB::ifPhysAddress.10 = STRING: a0:f8:49:f6:9e:5f
IF-MIB::ifPhysAddress.11 = STRING: a0:f8:49:f6:9e:60
IF-MIB::ifPhysAddress.12 = STRING: a0:f8:49:f6:9e:61
IF-MIB::ifPhysAddress.13 = STRING: a0:f8:49:f6:9e:62
IF-MIB::ifPhysAddress.14 = STRING: a0:f8:49:f6:9e:63
IF-MIB::ifPhysAddress.15 = STRING: a0:f8:49:f6:9e:64
IF-MIB::ifPhysAddress.16 = STRING: a0:f8:49:f6:9e:65
IF-MIB::ifPhysAddress.17 = STRING: a0:f8:49:f6:9e:66
IF-MIB::ifPhysAddress.18 = STRING: a0:f8:49:f6:9e:67
IF-MIB::ifPhysAddress.19 = STRING: a0:f8:49:f6:9e:68
IF-MIB::ifPhysAddress.20 = STRING: a0:f8:49:f6:9e:69
IF-MIB::ifPhysAddress.21 = STRING: a0:f8:49:f6:9e:6a
IF-MIB::ifPhysAddress.22 = STRING: a0:f8:49:f6:9e:6b
IF-MIB::ifPhysAddress.23 = STRING: a0:f8:49:f6:9e:6c
IF-MIB::ifPhysAddress.24 = STRING: a0:f8:49:f6:9e:6d
**IF-MIB::ifPhysAddress.25 = STRING: a0:f8:49:f6:9e:55
IF-MIB::ifPhysAddress.26 = STRING: a0:f8:49:f6:9e:55
IF-MIB::ifPhysAddress.27 = STRING: a0:f8:49:f6:9e:55
IF-MIB::ifPhysAddress.28 = STRING: a0:f8:49:f6:9e:55
IF-MIB::ifPhysAddress.29 = STRING: a0:f8:49:f6:9e:55
...
IF-MIB::ifPhysAddress.100201 = STRING: a0:f8:49:f6:9e:55
IF-MIB::ifPhysAddress.100202 = STRING: a0:f8:49:f6:9e:55
IF-MIB::ifPhysAddress.100203 = STRING: a0:f8:49:f6:9e:55
IF-MIB::ifPhysAddress.100305 = STRING: a0:f8:49:f6:9e:55**

**Comment by renatoalmeidaoliveira** [comment]
Hello @netravnen, @cguadall, @tabasco0322 I couldn't recreate that error in my environment so I increased the log level, could you please update the package to version 0.0.4 and post the error message?

**Comment by cguadall** [comment]
/usr/local/bin/snmp-diode -n redacted-ip -v 2 -c redacted-community
ERROR: The following errors were encountered during discovery:
ERROR: redacted-ip - 'ifPhysAddress'
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/snmp_diode/entrypoint.py", line 126, in main
    device_data = discover.gater_device_data(str(address), snmp_data, args.role, args.site)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/snmp_diode/discover.py", line 37, in gater_device_data
    interfaces = process_interfaces(session)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/snmp_diode/discover.py", line 69, in process_interfaces
    interfaces[interface_id]["mac"] = item.value.replace('"', "")[:-1].replace(
    ~~~~~~~~~~^^^^^^^^^^^^^^
KeyError: 'ifPhysAddress'

[]

**Comment by cguadall** [comment]
Test ran on an old Cisco Catalyst 3850 with the same error

Switch Ports Model              SW Version        SW Image              Mode
------ ----- -----              ----------        ----------            ----
     1 56    WS-C3850-48P       03.03.05.SE       cat3k_caa-universalk9 BUNDLE
*    2 56    WS-C3850-48P       03.03.05.SE       cat3k_caa-universalk9 BUNDLE


$ /usr/local/bin/snmp-diode -n redacted-ip -v 2 -c redacted-community
ERROR: The following errors were encountered during discovery:
ERROR: redacted-ip - 'ifPhysAddress'
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/snmp_diode/entrypoint.py", line 126, in main
    device_data = discover.gater_device_data(str(address), snmp_data, args.role, args.site)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/snmp_diode/discover.py", line 37, in gater_device_data
    interfaces = process_interfaces(session)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/snmp_diode/discover.py", line 69, in process_interfaces
    interfaces[interface_id]["mac"] = item.value.replace('"', "")[:-1].replace(
    ~~~~~~~~~~^^^^^^^^^^^^^^
KeyError: 'ifPhysAddress'

[]

**Comment by renatoalmeidaoliveira** [comment]
@cguadall, @netravnen, @tabasco0322 fixed, please update to 0.0.5.
